server.py

```diff
@@ -190,12 +190,29 @@ def announce():
 		server["total_clients"] = server["clients"]
 	server["pop_v"] = server["total_clients"] / server["updates"]
 
+	old_err = errorTracker.get(getErrorPK(server))
+
 	finishRequestAsync(server)
 
+	if old_err:
+		return ("Request has been filed, "
+			"but the previous request encountered the following error:\n" +
+			old_err, 409)
 	return "Request has been filed.", 202
 
 # Utilities
 
+# returns a primary key suitable for saving and replaying an error unique to a
+# server that was announced.
+def getErrorPK(server):
+	# We need to include the client IP in here, since some failures
+	# only happen depending on it.
+	return "%s/%s/%d" % (server["ip"], server["address"], server["port"])
+
+def isDomain(s):
+	# expressed as a regex: \.[A-Za-z][^.]*$
```

Comment:
Why not express it as a regex then?

Reply:
probably slower

Reply:
this particular regex should be linear time even with a backtracking regex engine like Python's

Reply:
```py
import re, timeit

def isDomain1(s):
	return "." in s and s.rpartition(".")[2][0].isalpha()

def isDomain2(s):
	i = s.rfind(".")
	j = len(s)
	return j >= 2 and 0 <= i <= j - 2 and s[i+1].isalpha()

def isDomain3(s):
	return re.search(r'\.[A-Za-z][^.]*$', s)

r_domain = re.compile(r'\.[A-Za-z][^.]*$')
def isDomain4(s):
	return r_domain.search(s)

print(timeit.timeit(lambda: isDomain1('test.com')))
print(timeit.timeit(lambda: isDomain2('test.com')))
print(timeit.timeit(lambda: isDomain3('test.com')))
print(timeit.timeit(lambda: isDomain4('test.com')))
```

results:
```
0.20050630900004762
0.24559447700085002
0.5673924059992714
0.2888703609969525
```

Reply:
If I'm reading that correctly, that's still only ca. 0.6 µs even for the slower "interpreted" variant. I don't think that's a problem.

> I only wrote the regex down to make it clearer what is actually checked

I think it would be most maintainable to just use the regex directly. But it doesn't matter terribly much.